### DISCUSSION TITLE
Heterogeneous workers on a cluster

### DISCUSSION THREAD

**Discussion by mivade** [dask/dask#7207] [Q&A]
Bioinformatics pipelines tend to consist of chaining together calls to various tools that are often written in compiled languages and have a command-line interface for specifying input and output files. Because each tool can require vastly different resources I would like to be able to submit work to a Dask cluster consisting of heterogeneous workers. I think this is relatively straightforward if using the command-line interface to start workers, e.g. (to take an example from the `distributed` docs):

```
dask-worker scheduler:8786 --resources "GPU=2"
dask-worker scheduler:8786 --resources "MEMORY=100e9"
```

and then I can specify what my required resources are when I submit a task:

```python
client.submit(my_task, arg, resources={"MEMORY": 70e9})
```

Ideally I wouldn't have to manually start the workers but could instead do this programmatically. Indeed it looks like this is exactly what `SpecCluster` allows me to do. Since the clusters provided by `dask_jobqueue` are now based on `SpecCluster` this seems like something I should be able to do, although if so it seems like the way to do this is pretty well obscured by the cluster implementations. Has anybody had any luck getting something like this to work?

Finally a quick digression: The (fairly small) overhead involved in starting Python processes to mostly call subprocesses is noted. The main advantage to me for using Dask for a bioinformatics workload is freeing myself from more cumbersome (IMHO) DSL-based tools like Snakemake, Nextflow, CWL, WDL, etc. and instead being able to mostly write code entirely in Python.

**Comment by SultanOrazbayev** [comment]
I tried moving my analysis from snakemake to dask/prefect, but so far without much luck. I'm curious to learn what others are doing. The basic problem I have is that in HPC environment queues don't guarantee simultaneous availability of resources, so it's not uncommon for snakemake to wait a couple of days until a specific compute node is available. Also, implementing file-based workflow in dask and prefect is a bit cumbersome (or maybe I was doing it wrong).

**Reply by mivade** [reply]
For what it's worth I have tried one approach that has had some mixed success: use separate Dask clusters for distinct steps. This makes things a bit harder to coordinate though because you either have to coordinate multiple Dask clusters at once or you have to explicitly wait on steps to finish. So something (very roughly) like this:

```python
# Generate FASTQ files
with HTCondorCluster(...) as cluster:
    with Client(cluster) as client:
        wait(client.submit(bcl2fastq, *args))

# Do subsequent analysis steps which may be able to take advantage of more workers
with HTCondorCluster(...) as cluster:
    cluster.adapt(**adapt_kwargs)

    with Client(cluster) as client:
        futures = [client.submit(job, arg) for arg in args]
        wait(futures)
```

File-based workflows can also be a bit tricky to handle depending on what you want to do. Typically I try to write tasks that operate on files to take in a path to a file as an argument and return a path to whatever result file it produces (this naturally fits in well with Dask if you have a single client because it knows what to do with futures as arguments). If you care about caching intermediate steps you have to handle this on your own but it shouldn't be too difficult to write a simple decorator or context manager to handle that. I tend to not worry about being able to pick up a failed workflow from an intermediate step since in a lot of my pipeline use cases it normally doesn't take too long to restart from the beginning. At the end of the day tools like Snakemake are probably better for these kinds of workflows although it takes a lot of discipline to create them in a maintainable way (and don't get me started on testability).
